### DISCUSSION TITLE
Speed up Dockerfile build time

### DISCUSSION THREAD

**Discussion by useragent7** [eclipse-emfcloud/emfcloud#172] [Coffee Editor] [answered]
Hi,
I wanted to play around with the coffee-editor and noticed that the creation of the docker images takes about 30 minutes on my machine.
I'm not that proficient with java so I wanted to ask if you tried to cache the building artifacts.
If not I would give it a shot.
My host is sadly on windows and because of the dependencies I can only run it via docker.

**Comment by lucas-koehler** [comment, marked as answer]
Hello @useragent7,
as you mentioned Java, I assume you mean the built artifacts from the `backend` folder, right?
The docker build currently does not cache any build artifacts, neither from the backend nor the frontend.
This is the case because the backend and frontend are all built in one [single step](https://github.com/eclipse-emfcloud/coffee-editor/blob/8661ae1c60d97d128f2fe3c68a0c9dd5185701c7/Dockerfile#L26).
Because the whole coffee-editor workspace is copied before (see [here](https://github.com/eclipse-emfcloud/coffee-editor/blob/8661ae1c60d97d128f2fe3c68a0c9dd5185701c7/Dockerfile#L23)), all following steps of the docker build have to be re-executed when ANY file in your workspace changes that is not excluded by the `.dockerignore` file (this is called cached invalidation in case you want to look it up).

An approach to improve caching could be the following:

1. Copy only the files needed for the backend build, i.e. the `backend` folder and the `run.sh` script
2. build the backend with `./run.sh -b`
3. Copy files needed for the frontend build
4. copy backend and build frontend with `./run.sh -cf`

continue with the rest of the docker build. With that the backend build does not need to be executed again as long as the contents of the `backend` folder do not change.

See also the docker documentation on best practices for docker files: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/

If you optimize the `Dockerfile`, gladly open a pull request so we can integrate the improvements :)
Best regards,
Lucas
